modules/messages/src/lanes_manager.rs

```diff
@@ -0,0 +1,280 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+	BridgedChainOf, Config, Error, InboundLane, InboundLaneStorage, InboundLanes, OutboundLane,
+	OutboundLaneStorage, OutboundLanes, OutboundMessages, StoredInboundLaneData,
+	StoredMessagePayload,
+};
+
+use bp_messages::{
+	ChainWithMessages, InboundLaneData, LaneId, LaneState, MessageKey, MessageNonce,
+	MessagePayload, OutboundLaneData, VerificationError,
+};
+use bp_runtime::AccountIdOf;
+use codec::MaxEncodedLen;
+use frame_support::{ensure, RuntimeDebug};
+use sp_std::marker::PhantomData;
+
+/// Lanes manager errors.
+#[derive(RuntimeDebug, PartialEq, Eq)]
+pub enum LanesManagerError {
+	/// Inbound lane already exists.
+	InboundLaneAlreadyExists,
+	/// Outbound lane already exists.
+	OutboundLaneAlreadyExists,
+	/// No inbound lane with given id.
+	UnknownInboundLane,
+	/// No outbound lane with given id.
+	UnknownOutboundLane,
+	/// Inbound lane with given id is closed.
+	ClosedInboundLane,
+	/// Outbound lane with given id is closed.
+	ClosedOutboundLane,
+}
+
+impl<T, I> From<LanesManagerError> for Error<T, I> {
+	fn from(e: LanesManagerError) -> Self {
+		match e {
+			LanesManagerError::InboundLaneAlreadyExists => Error::InboundLaneAlreadyExists,
+			LanesManagerError::OutboundLaneAlreadyExists => Error::OutboundLaneAlreadyExists,
+			LanesManagerError::UnknownInboundLane => Error::UnknownInboundLane,
+			LanesManagerError::UnknownOutboundLane => Error::UnknownOutboundLane,
+			LanesManagerError::ClosedInboundLane => Error::ClosedInboundLane,
+			LanesManagerError::ClosedOutboundLane => Error::ClosedOutboundLane,
+		}
+	}
+}
+
+/// Message lanes manager.
+pub struct LanesManager<T, I>(PhantomData<(T, I)>);
+
+impl<T: Config<I>, I: 'static> Default for LanesManager<T, I> {
+	fn default() -> Self {
+		Self::new()
+	}
+}
+
+impl<T: Config<I>, I: 'static> LanesManager<T, I> {
+	/// Create new lanes manager.
+	pub fn new() -> Self {
```

Comment:
Not sure, but imo Rust is smart enough and there's no any overhead for zero-byte instances. I've made it instance-level methods, because in other pallet I'm calling multiple methods of the lanes manager from the same function and imo it is easier to read
```rust
let lanes_manager = LanesManagerOf::<T, I>::new();
...
lanes_manage.create_inbound_lane();
...
lanes_manage.create_outbound_lane();
```

than
```rust
LanesManagerOf::<T, I>::create_inbound_lane();
...
LanesManagerOf::<T, I>::create_outbound_lane();
```

But that's just my personal feeling - feel free to insist if you think it needs to be fixed

Reply:
Oh, ok. No, I don't have strong preferences. Looks good.